unity-renderer/Assets/Scripts/MainScripts/DCL/AvatarSystem/Loader/SceneEmoteHelper.cs

```diff
@@ -0,0 +1,97 @@
+using Cysharp.Threading.Tasks;
+using DCL.Controllers;
+using DCL.Emotes;
+using System;
+using System.Collections.Generic;
+using System.Threading;
+
+namespace AvatarSystem
+{
+    public static class SceneEmoteHelper
+    {
+        public const string SCENE_EMOTE_PREFIX = "urn:decentraland:off-chain:scene-emote:";
+
+        public static bool TryGenerateEmoteId(IParcelScene scene, string emoteFilePath, bool loop, out string emoteId)
+        {
+            if (scene.contentProvider.TryGetContentHash(emoteFilePath, out string emoteHash))
+            {
+                emoteId = $"{SCENE_EMOTE_PREFIX}{emoteHash}-{(loop ? "true" : "false")}";
+                return true;
+            }
+
+            emoteId = string.Empty;
+            return false;
+        }
+
+        public static bool TryGetDataFromEmoteId(string emoteId, out string emoteHash, out bool loop)
+        {
+            try
+            {
+                ReadOnlySpan<char> prefixRemoved = emoteId.AsSpan().Slice(SCENE_EMOTE_PREFIX.Length, emoteId.Length - SCENE_EMOTE_PREFIX.Length);
+                int loopSeparator = prefixRemoved.LastIndexOf('-');
+                emoteHash = prefixRemoved.Slice(0, loopSeparator).ToString();
+
+                var loopSpan = prefixRemoved.Slice(loopSeparator + 1, 4);
+                loop = loopSpan.Equals("true", StringComparison.InvariantCultureIgnoreCase);
+
+                return true;
+            }
+            catch (Exception _)
+            {
+                emoteHash = string.Empty;
+                loop = false;
+            }
+
+            return false;
+        }
+
+        public static bool IsSceneEmote(string emoteId)
+        {
+            return emoteId.StartsWith(SCENE_EMOTE_PREFIX);
+        }
+
+        public static async UniTask RequestLoadSceneEmote(
```

Comment:
I recommend considerably flattening and simplifying the flow:
- You have many collections that serve a single purpose: ensuring that the flow for a given `(string bodyshapeId, string emoteId)` is launched only once. You can use a single collection with `UniTaskCompletionSource`, we have plenty of examples in our codebase of how to use it. TLDR; you should produce `UniTasks` from the cached source without any branching or reliance on third-party collections. If the result is already evaluated `UniTaskCompletionSource` can give it straight away so you don't need to branch your logic.
- `Func<bool> cancelCondition` looks redundant, there is a common mechanism of[ linking cancellation tokens](https://learn.microsoft.com/en-us/dotnet/api/system.threading.cancellationtokensource.createlinkedtokensource?view=net-7.0) so you simply create a token that will react on both local and disposal cancellation.

Reply:
replaced that lambda for `CancellationTokenSource` here https://github.com/decentraland/unity-renderer/pull/5501/commits/04a3ef16362e9dda561cbc2cda137ea51def5979
regarding `UniTaskCompletionSource` I'm not sure, because we will be generating extra GC for every new request since `UniTaskCompletionSource` it's a class and can't be reset so we can't even use a object pool.
I agree that handling two hashsets for pending/completed is not ideal, but I don't believe it's complex enough to worth all those extra allocations.
If you have any ideas to solve that situation, I'm all ears